crates/jstzd/src/protocol/bootstrap.rs

```diff
@@ -0,0 +1,157 @@
+use std::collections::HashSet;
+
+use jstz_crypto::public_key::PublicKey;
+use serde_json::Value;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct BootstrapAccount {
+    public_key: PublicKey,
+    amount_mutez: u64,
+}
+
+impl TryFrom<&Value> for BootstrapAccount {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &Value) -> Result<Self, Self::Error> {
+        let array = value
+            .as_array()
+            .ok_or(anyhow::anyhow!("value is not a valid json array"))?;
+        if array.len() != 2 {
+            return Err(anyhow::anyhow!(
+                "value is not in the acceptable format for bootstrap accounts"
+            ));
+        }
+        let public_key = array
+            .first()
+            .unwrap()
+            .as_str()
+            .ok_or(anyhow::anyhow!("'public_key' is not a valid string"))?;
+        let amount_mutez = array
+            .get(1)
+            .unwrap()
+            .as_str()
+            .ok_or(anyhow::anyhow!(
+                "'amount' is not a valid string representing an unsigned integer"
+            ))?
+            .parse::<u64>();
+        if amount_mutez.is_err() {
+            return Err(anyhow::anyhow!(
+                "'amount' is not a valid string representing an unsigned integer"
+            ));
+        }
+        Ok(Self {
+            public_key: PublicKey::from_base58(public_key)?,
+            amount_mutez: amount_mutez.unwrap(),
+        })
+    }
+}
+
+impl From<&BootstrapAccount> for Value {
+    fn from(value: &BootstrapAccount) -> Self {
+        Value::Array(vec![
+            Value::String(value.public_key.to_string()),
+            Value::String(value.amount_mutez.to_string()),
+        ])
+    }
+}
+
+impl BootstrapAccount {
+    pub fn new(public_key: &str, amount_mutez: u64) -> anyhow::Result<Self> {
+        Ok(Self {
+            public_key: PublicKey::from_base58(public_key)?,
+            amount_mutez,
+        })
+    }
+
+    #[cfg(test)]
+    pub fn amount(&self) -> u64 {
+        self.amount_mutez
+    }
+}
+
+#[derive(Default)]
+pub struct BootstrapAccounts {
+    keys: HashSet<String>,
+    accounts: Vec<BootstrapAccount>,
```

Comment:
HashMap? or `IndexMap` if you want to preserve the insertion ordering or `BTreeMap` if you want to visit the keys in sorted order.

Reply:
Changed to HashMap. The method `accounts` gets more costly but it should be fine since it's not called often.